### ISSUE TITLE
2023.12.3 - Unable to subset by multiple items in searchString with Herbie.xarray, but works for Herbie.inventory

### ISSUE THREAD

**Issue by fleegs79** [blaylockbk/Herbie#259]
Hello,

With 2023.12.3, when I try to download a subset of data based on multiple items in a `searchString`, it will download the entire dataset, instead of just the items requested. If I roll back to 2023.12.2 or 2023.12.1, the multiple item `searchString` and subsetting works as expected.  Also, if I provide a single item to the `searchString` (ex: `":ASNOW:"`) it subsets the data fine.  

Here's an example (using Google Colab): 

```python
model = 'rap'
product = 'awp130pgrb'
dtg = '2023-12-25 21:00'
fhr = 0
H = Herbie(dtg, model=Model, product=product, fxx=fhr)
print (H.inventory(searchString="GUST|UGRD:10 m|VGRD:10 m|TMP:2 m|DPT:2 m"))
```

produces: 
```
     grib_message  start_byte    end_byte              range  \
5             6.0      493398    568114.0      493398-568114   
239         203.0    10920489  10990621.0  10920489-10990621   
242         206.0    11149757  11223824.0  11149757-11223824   
247         211.1    11480445  11480444.0  11480445-11480444   
248         211.2    11480445  11615790.0  11480445-11615790   

         reference_time          valid_time variable              level  \
5   2023-12-25 21:00:00 2023-12-25 21:00:00     GUST            surface   
239 2023-12-25 21:00:00 2023-12-25 21:00:00      TMP   2 m above ground   
242 2023-12-25 21:00:00 2023-12-25 21:00:00      DPT   2 m above ground   
247 2023-12-25 21:00:00 2023-12-25 21:00:00     UGRD  10 m above ground   
248 2023-12-25 21:00:00 2023-12-25 21:00:00     VGRD  10 m above ground   

    forecast_time                  search_this  
5             anl            :GUST:surface:anl  
239           anl    :TMP:2 m above ground:anl  
242           anl    :DPT:2 m above ground:anl  
247           anl  :UGRD:10 m above ground:anl  
248           anl  :VGRD:10 m above ground:anl  
```

and matches my expectations.  But when using this line:
```python
data = H.xarray(searchString="GUST|UGRD:10 m|VGRD:10 m|TMP:2 m|DPT:2 m")
data
```

The first 5 curl lines are quickly listed and match the start/end bytes listed above,  
```
curl -s --range 493398-568114 "https://noaa-rap-pds.s3.amazonaws.com/rap.20231225/rap.t21z.awp130pgrbf00.grib2" > "/root/data/rap/20231225/subset_38ef7551__rap.t21z.awp130pgrbf00.grib2"
curl -s --range 10920489-10990621 "https://noaa-rap-pds.s3.amazonaws.com/rap.20231225/rap.t21z.awp130pgrbf00.grib2" >> "/root/data/rap/20231225/subset_38ef7551__rap.t21z.awp130pgrbf00.grib2"
curl -s --range 11149757-11223824 "https://noaa-rap-pds.s3.amazonaws.com/rap.20231225/rap.t21z.awp130pgrbf00.grib2" >> "/root/data/rap/20231225/subset_38ef7551__rap.t21z.awp130pgrbf00.grib2"
curl -s --range 11480445-11480444 "https://noaa-rap-pds.s3.amazonaws.com/rap.20231225/rap.t21z.awp130pgrbf00.grib2" >> "/root/data/rap/20231225/subset_38ef7551__rap.t21z.awp130pgrbf00.grib2"
curl -s --range 11480445-11615790 "https://noaa-rap-pds.s3.amazonaws.com/rap.20231225/rap.t21z.awp130pgrbf00.grib2" >> "/root/data/rap/20231225/subset_38ef7551__rap.t21z.awp130pgrbf00.grib2"
```

then it spends a while processing data before providing the following:
```
`Note: Returning a list of [39] xarray.Datasets because cfgrib opened with multiple hypercubes.`
[<xarray.Dataset>
 Dimensions:              (y: 337, x: 451)
 Coordinates:
     time                 datetime64[ns] 2023-12-25T21:00:00
     step                 timedelta64[ns] 00:00:00
     atmosphere           float64 0.0
     latitude             (y, x) float64 16.28 16.31 16.34 ... 55.54 55.51 55.48
     longitude            (y, x) float64 233.9 234.0 234.1 ... 302.3 302.4 302.6
     valid_time           datetime64[ns] 2023-12-25T21:00:00
 Dimensions without coordinates: y, x
 Data variables:
     unknown              (y, x) float32 61.0 61.0 61.0 62.0 ... 83.0 83.0 83.0
     tcc                  (y, x) float32 100.0 100.0 100.0 ... 100.0 100.0 100.0
     refc                 (y, x) float32 0.5625 -4.125 -1.688 ... 8.5 8.5 5.375
     gribfile_projection  object None
 Attributes:
     GRIB_edition:            2
     GRIB_centre:             kwbc
     GRIB_centreDescription:  US National Weather Service - NCEP
     GRIB_subCentre:          0
     Conventions:             CF-1.7
     institution:             US National Weather Service - NCEP
     model:                   rap
     product:                 awp130pgrb
     description:             Rapid Refresh (RAP) from NOMADS and Big Data Pro...
     remote_grib:             https://noaa-rap-pds.s3.amazonaws.com/rap.202312...
     local_grib:              /root/data/rap/20231225/subset_38ef7551__rap.t21...
     searchString:            GUST|UGRD:10 m|VGRD:10 m|TMP:2 m|DPT:2 m,
 <xarray.Dataset>
 Dimensions:                (y: 337, x: 451)
 Coordinates:
     time                   datetime64[ns] 2023-12-25T21:00:00
     step                   timedelta64[ns] 00:00:00
     atmosphereSingleLayer  float64 0.0
     latitude               (y, x) float64 16.28 16.31 16.34 ... 55.51 55.48
     longitude              (y, x) float64 233.9 234.0 234.1 ... 302.4 302.6
     valid_time             datetime64[ns] 2023-12-25T21:00:00
 Dimensions without coordinates: y, x
 Data variables:
     unknown                (y, x) float32 -5e+03 -5e+03 ... 5.625e+03 2.723e+03
     pwat                   (y, x) float32 36.22 35.97 36.1 ... 13.47 13.35 13.22
     gribfile_projection    object None
 Attributes:
     GRIB_edition:            2
     GRIB_centre:             kwbc
     GRIB_centreDescription:  US National Weather Service - NCEP
     GRIB_subCentre:          0
     Conventions:             CF-1.7
     institution:             US National Weather Service - NCEP
     model:                   rap
     product:                 awp130pgrb
     description:             Rapid Refresh (RAP) from NOMADS and Big Data Pro...
     remote_grib:             https://noaa-rap-pds.s3.amazonaws.com/rap.202312...
     local_grib:              /root/data/rap/20231225/subset_38ef7551__rap.t21...
     searchString:            GUST|UGRD:10 m|VGRD:10 m|TMP:2 m|DPT:2 m,
 <xarray.Dataset>
 Dimensions:                  (y: 337, x: 451)
 Coordinates:
     time                     datetime64[ns] 2023-12-25T21:00:00
     step                     timedelta64[ns] 00:00:00
     boundaryLayerCloudLayer  float64 0.0
     latitude                 (y, x) float64 16.28 16.31 16.34 ... 55.51 55.48
     longitude                (y, x) float64 233.9 234.0 234.1 ... 302.4 302.6
     valid_time               datetime64[ns] 2023-12-25T21:00:00
 Dimensions without coordinates: y, x
 Data variables:
     tcc                      (y, x) float32 100.0 100.0 100.0 ... 100.0 100.0
     gribfile_projection      object None
 Attributes:
     GRIB_edition:            2
     GRIB_centre:             kwbc
     GRIB_centreDescription:  US National Weather Service - NCEP
     GRIB_subCentre:          0
     Conventions:             CF-1.7
     institution:             US National Weather Service - NCEP
     model:                   rap
     product:                 awp130pgrb
     description:             Rapid Refresh (RAP) from NOMADS and Big Data Pro...
     remote_grib:             https://noaa-rap-pds.s3.amazonaws.com/rap.202312...
     local_grib:              /root/data/rap/20231225/subset_38ef7551__rap.t21...
     searchString:            GUST|UGRD:10 m|VGRD:10 m|TMP:2 m|DPT:2 m,
 <xarray.Dataset>
 Dimensions:              (y: 337, x: 451)
 Coordinates:
     time                 datetime64[ns] 2023-12-25T21:00:00
     step                 timedelta64[ns] 00:00:00
     cloudBase            float64 0.0
     latitude             (y, x) float64 16.28 16.31 16.34 ... 55.54 55.51 55.48
     longitude            (y, x) float64 233.9 234.0 234.1 ... 302.3 302.4 302.6
     valid_time           datetime64[ns] 2023-12-25T21:00:00
 Dimensions without coordinates: y, x
 Data variables:
     gh                   (y, x) float32 747.0 781.0 814.5 ... 678.0 677.0 676.0
     gribfile_projection  object None
 Attributes:
     GRIB_edition:            2
     GRIB_centre:             kwbc
     GRIB_centreDescription:  US National Weather Service - NCEP
     GRIB_subCentre:          0
     Conventions:             CF-1.7
     institution:             US National Weather Service - NCEP
     model:                   rap
     product:                 awp130pgrb
     description:             Rapid Refresh (RAP) from NOMADS and Big Data Pro...
     remote_grib:             https://noaa-rap-pds.s3.amazonaws.com/rap.202312...
     local_grib:              /root/data/rap/20231225/subset_38ef7551__rap.t21...
     searchString:            GUST|UGRD:10 m|VGRD:10 m|TMP:2 m|DPT:2 m,
```
This is only a small subset of the data included in the "data" xarray, but it contains information that wasn't requested with the searchString.  

The same thing happens with this searchString: 
`"((HGT|UGRD|VGRD|TMP):((850|500|250) mb))"`

Am I using improper coding for the `searchString` now in 2023.12.3 (compared to the previous versions)? Any other ideas on why this is occurring? 

Thanks in advance for the help and I appreciate all of your work on Herbie. It's been a huge help for me to grab past model data. 

Steve

**Comment by blaylockbk** [comment]
Thanks for reporting this issue. I'll take a look.

**Comment by blaylockbk** [comment]
So far in my investigation, this only seems to be an issue with the RAP model. I'll keep looking into why.

**Comment by blaylockbk** [comment]
Ah, the issue is caused by the grib submessages.

<img width="623" alt="image" src="https://github.com/blaylockbk/Herbie/assets/6249613/85c7e673-cfa7-4edf-bde3-0dc4afd6c6ef">

The UGRD and VGRD messages are packed together, and the curl Range is incorrect.

I'll need to update this bug to only download messages if the curl range is valid.

**Comment by blaylockbk** [comment]
Thanks @fleegs79 for reporting this issue so quickly 😁 I hope the quick fix in #260 will prevent this from ever happening. Give `2023.12.4` a try and please report any other issues.

**Comment by fleegs79** [comment]
Thank you for troubleshooting and the quick fix!  I tested it this evening and it works as expected.  Thanks again! 

-Steve
